sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidPlanner.java

```diff
@@ -761,84 +787,204 @@ public T next()
-          query = new SqlOrderBy(
-              query.getParserPosition(),
-              query,
-              druidSqlInsert.getClusteredBy(),
-              offset,
-              fetch
-          );
+
+          SqlNode replaceTimeQuery = druidSqlReplace.getReplaceTimeQuery();
+          if (replaceTimeQuery == null) {
+            throw new ValidationException("Missing time chunk information in DELETE WHERE clause for replace.");
+          }
+
+          ingestionGranularity = druidSqlReplace.getPartitionedBy();
+          replaceIntervals = validateQueryAndConvertToIntervals(replaceTimeQuery, ingestionGranularity, dateTimeZone);
         }
       }
 
       if (!query.isA(SqlKind.QUERY)) {
         throw new ValidationException(StringUtils.format("Cannot execute [%s].", query.getKind()));
       }
 
-      return new ParsedNodes(explain, druidSqlInsert, query, ingestionGranularity);
+      if (druidSqlInsert != null) {
+        return new ParsedNodes(explain, druidSqlInsert, query, ingestionGranularity, null);
+      } else {
+        return new ParsedNodes(explain, druidSqlReplace, query, ingestionGranularity, replaceIntervals);
+      }
+    }
+
+    private static List<String> validateQueryAndConvertToIntervals(
+        SqlNode replaceTimeQuery,
+        Granularity granularity,
+        DateTimeZone dateTimeZone
+    ) throws ValidationException
+    {
+      if (replaceTimeQuery instanceof SqlLiteral && "all".equalsIgnoreCase(((SqlLiteral) replaceTimeQuery).toValue())) {
+        return ImmutableList.of("all");
+      }
+
+      DimFilter dimFilter = convertQueryToDimFilter(replaceTimeQuery, dateTimeZone);
+      if (!ImmutableSet.of(ColumnHolder.TIME_COLUMN_NAME).equals(dimFilter.getRequiredColumns())) {
+        throw new ValidationException("Only " + ColumnHolder.TIME_COLUMN_NAME + " column is supported in DELETE WHERE clause");
+      }
+
+      Filtration filtration = Filtration.create(dimFilter);
+      filtration = MoveTimeFiltersToIntervals.instance().apply(filtration);
+      List<Interval> intervals = filtration.getIntervals();
+
+      for (Interval interval : intervals) {
+        DateTime intervalStart = interval.getStart();
+        DateTime intervalEnd = interval.getEnd();
+        // The start of each interval should be aligned with the start of the bucket, and the end interval should be
+        // aligned with the start of the next interval.
+        if (!granularity.bucketStart(intervalStart).equals(intervalStart) || !granularity.bucketStart(intervalEnd).equals(intervalEnd)) {
+          throw new ValidationException("DELETE WHERE clause contains an interval which is not aligned with PARTITIONED BY granularity");
+        }
+      }
+      return intervals
+          .stream()
+          .map(AbstractInterval::toString)
+          .collect(Collectors.toList());
+    }
+
+    private static DimFilter convertQueryToDimFilter(SqlNode replaceTimeQuery, DateTimeZone dateTimeZone)
```

Comment:
We should move this to `DruidSqlParserUtils`. Similar comment for all the other functions.

Reply:
Moved